server: TestCachedSettingsServerRestart failed

server.TestCachedSettingsServerRestart [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/15331087?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/15331087?buildTab=artifacts#/) on release-23.1 @ [d9b0e5f8cefa99bdcc217f6be790d424c603031c](https://github.com/cockroachdb/cockroach/commits/d9b0e5f8cefa99bdcc217f6be790d424c603031c):


```
=== RUN   TestCachedSettingsServerRestart
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestCachedSettingsServerRestart1780436221
    test_log_scope.go:79: use -show-logs to present logs inline
    settings_cache_test.go:141: condition failed to evaluate within 3m45s: initial state settings KVs does not match expected settings
        Expected: [{Key:/Table/6/1/"diagnostics.reporting.enabled"/0 Value:{RawBytes:[186 121 60 169 10 38 4 116 114 117 101 24 178 155 218 228 12 160 211 211 158 2 22 1 98] Timestamp:0,0}} {Key:/Table/6/1/"version"/0 Value:{RawBytes:[245 212 250 113 10 38 10 18 8 8 23 16 1 24 0 32 0 24 178 155 218 228 12 144 162 253 235 3 22 1 109] Timestamp:0,0}}]
        Actual:   [{Key:/Table/6/1/"cluster.secret"/0 Value:{RawBytes:[167 136 241 49 10 38 36 100 99 99 49 49 57 51 100 45 49 52 98 52 45 52 49 56 50 45 98 98 55 49 45 57 102 98 56 99 99 100 54 49 54 56 97 24 178 155 218 228 12 144 225 129 180 5 22 1 115] Timestamp:0,0}} {Key:/Table/6/1/"diagnostics.reporting.enabled"/0 Value:{RawBytes:[186 121 60 169 10 38 4 116 114 117 101 24 178 155 218 228 12 160 211 211 158 2 22 1 98] Timestamp:0,0}} {Key:/Table/6/1/"version"/0 Value:{RawBytes:[245 212 250 113 10 38 10 18 8 8 23 16 1 24 0 32 0 24 178 155 218 228 12 144 162 253 235 3 22 1 109] Timestamp:0,0}}]
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestCachedSettingsServerRestart1780436221
--- FAIL: TestCachedSettingsServerRestart (230.54s)
```

Parameters:
 - <code>TAGS=bazel,gss,race</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCachedSettingsServerRestart.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-38882

server.TestCachedSettingsServerRestart [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/15814830?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/15814830?buildTab=artifacts#/) on release-23.1 @ [4534017ff77b216dcad4301f13d8ee13cf7fd423](https://github.com/cockroachdb/cockroach/commits/4534017ff77b216dcad4301f13d8ee13cf7fd423):


```
=== RUN   TestCachedSettingsServerRestart
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestCachedSettingsServerRestart3791630463
    test_log_scope.go:79: use -show-logs to present logs inline
    settings_cache_test.go:141: condition failed to evaluate within 3m45s: initial state settings KVs does not match expected settings
        Expected: [{Key:/Table/6/1/"diagnostics.reporting.enabled"/0 Value:{RawBytes:[90 19 170 150 10 38 4 116 114 117 101 24 128 138 214 231 12 208 237 197 155 3 22 1 98] Timestamp:0,0}} {Key:/Table/6/1/"version"/0 Value:{RawBytes:[24 40 96 157 10 38 10 18 8 8 23 16 1 24 0 32 0 24 128 138 214 231 12 208 165 130 224 4 22 1 109] Timestamp:0,0}}]
        Actual:   [{Key:/Table/6/1/"cluster.secret"/0 Value:{RawBytes:[28 1 185 15 10 38 36 50 101 54 100 56 56 48 97 45 97 50 55 98 45 52 52 55 53 45 56 54 52 49 45 99 51 54 50 98 51 57 57 54 49 102 51 24 128 138 214 231 12 176 160 233 153 6 22 1 115] Timestamp:0,0}} {Key:/Table/6/1/"diagnostics.reporting.enabled"/0 Value:{RawBytes:[90 19 170 150 10 38 4 116 114 117 101 24 128 138 214 231 12 208 237 197 155 3 22 1 98] Timestamp:0,0}} {Key:/Table/6/1/"version"/0 Value:{RawBytes:[24 40 96 157 10 38 10 18 8 8 23 16 1 24 0 32 0 24 128 138 214 231 12 208 165 130 224 4 22 1 109] Timestamp:0,0}}]
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestCachedSettingsServerRestart3791630463
--- FAIL: TestCachedSettingsServerRestart (230.37s)
```

Parameters:
 - <code>TAGS=bazel,gss,race</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #125429 server: TestCachedSettingsServerRestart failed [C-test-failure O-robot T-server-and-security branch-release-23.1.23-rc release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCachedSettingsServerRestart.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestCachedSettingsServerRestart [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/15870005?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/15870005?buildTab=artifacts#/) on release-23.1 @ [fbcb992a72c3ac2a9af96f6238a24b3978bcaadf](https://github.com/cockroachdb/cockroach/commits/fbcb992a72c3ac2a9af96f6238a24b3978bcaadf):


```
=== RUN   TestCachedSettingsServerRestart
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestCachedSettingsServerRestart3168928694
    test_log_scope.go:79: use -show-logs to present logs inline
    settings_cache_test.go:141: condition failed to evaluate within 3m45s: initial state settings KVs does not match expected settings
        Expected: [{Key:/Table/6/1/"diagnostics.reporting.enabled"/0 Value:{RawBytes:[251 2 248 210 10 38 4 116 114 117 101 24 212 168 128 232 12 160 253 155 157 7 22 1 98] Timestamp:0,0}}]
        Actual:   [{Key:/Table/6/1/"diagnostics.reporting.enabled"/0 Value:{RawBytes:[251 2 248 210 10 38 4 116 114 117 101 24 212 168 128 232 12 160 253 155 157 7 22 1 98] Timestamp:0,0}} {Key:/Table/6/1/"version"/0 Value:{RawBytes:[151 128 229 124 10 38 10 18 8 8 23 16 1 24 0 32 0 24 214 168 128 232 12 240 186 254 165 1 22 1 109] Timestamp:0,0}}]
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestCachedSettingsServerRestart3168928694
--- FAIL: TestCachedSettingsServerRestart (230.01s)
```

Parameters:
 - <code>TAGS=bazel,gss,race</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #125429 server: TestCachedSettingsServerRestart failed [C-test-failure O-robot T-server-and-security branch-release-23.1.23-rc release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCachedSettingsServerRestart.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestCachedSettingsServerRestart [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/16194387?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/16194387?buildTab=artifacts#/) on release-23.1 @ [2073003f4a56bcb7eba3e76deae2df14151ed137](https://github.com/cockroachdb/cockroach/commits/2073003f4a56bcb7eba3e76deae2df14151ed137):


```
=== RUN   TestCachedSettingsServerRestart
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestCachedSettingsServerRestart3441775381
    test_log_scope.go:79: use -show-logs to present logs inline
    settings_cache_test.go:141: condition failed to evaluate within 3m45s: initial state settings KVs does not match expected settings
        Expected: [{Key:/Table/6/1/"diagnostics.reporting.enabled"/0 Value:{RawBytes:[111 25 204 142 10 38 4 116 114 117 101 24 246 182 253 233 12 144 146 222 17 22 1 98] Timestamp:0,0}}]
        Actual:   [{Key:/Table/6/1/"diagnostics.reporting.enabled"/0 Value:{RawBytes:[111 25 204 142 10 38 4 116 114 117 101 24 246 182 253 233 12 144 146 222 17 22 1 98] Timestamp:0,0}} {Key:/Table/6/1/"version"/0 Value:{RawBytes:[81 119 54 33 10 38 10 18 8 8 23 16 1 24 0 32 0 24 246 182 253 233 12 176 137 192 212 1 22 1 109] Timestamp:0,0}}]
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestCachedSettingsServerRestart3441775381
--- FAIL: TestCachedSettingsServerRestart (230.14s)
```

Parameters:
 - <code>TAGS=bazel,gss,race</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #125429 server: TestCachedSettingsServerRestart failed [C-test-failure O-robot T-server-and-security branch-release-23.1.23-rc release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCachedSettingsServerRestart.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestCachedSettingsServerRestart [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/16346955?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/16346955?buildTab=artifacts#/) on release-23.1 @ [7748ab2d671c6e8d021af1ca577d2de4578751db](https://github.com/cockroachdb/cockroach/commits/7748ab2d671c6e8d021af1ca577d2de4578751db):


```
=== RUN   TestCachedSettingsServerRestart
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestCachedSettingsServerRestart3485642125
    test_log_scope.go:79: use -show-logs to present logs inline
    settings_cache_test.go:141: condition failed to evaluate within 3m45s: initial state settings KVs does not match expected settings
        Expected: [{Key:/Table/6/1/"diagnostics.reporting.enabled"/0 Value:{RawBytes:[77 95 233 0 10 38 4 116 114 117 101 24 226 139 145 235 12 128 231 132 128 5 22 1 98] Timestamp:0,0}} {Key:/Table/6/1/"version"/0 Value:{RawBytes:[205 190 79 226 10 38 10 18 8 8 23 16 1 24 0 32 0 24 226 139 145 235 12 176 197 172 197 6 22 1 109] Timestamp:0,0}}]
        Actual:   [{Key:/Table/6/1/"cluster.secret"/0 Value:{RawBytes:[219 97 94 49 10 38 36 100 52 54 49 53 48 97 57 45 56 48 102 56 45 52 52 97 57 45 57 53 102 49 45 97 51 52 50 102 57 53 102 55 52 55 51 24 228 139 145 235 12 208 171 233 73 22 1 115] Timestamp:0,0}} {Key:/Table/6/1/"diagnostics.reporting.enabled"/0 Value:{RawBytes:[77 95 233 0 10 38 4 116 114 117 101 24 226 139 145 235 12 128 231 132 128 5 22 1 98] Timestamp:0,0}} {Key:/Table/6/1/"version"/0 Value:{RawBytes:[205 190 79 226 10 38 10 18 8 8 23 16 1 24 0 32 0 24 226 139 145 235 12 176 197 172 197 6 22 1 109] Timestamp:0,0}}]
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestCachedSettingsServerRestart3485642125
--- FAIL: TestCachedSettingsServerRestart (230.47s)
```

Parameters:
 - <code>TAGS=bazel,gss,race</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #125429 server: TestCachedSettingsServerRestart failed [C-test-failure O-robot T-product-security branch-release-23.1.23-rc release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCachedSettingsServerRestart.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/server/settings_cache_test.go
+++ b/pkg/server/settings_cache_test.go
@@ -77,7 +77,7 @@ func TestCachedSettingsServerRestart(t *testing.T) {
 			},
 		},
 	}
-	var settingsCache []roachpb.KeyValue
+	var expectedSettingsCache []roachpb.KeyValue
 	testServer, _, _ := serverutils.StartServer(t, serverArgs)
 
 	factor := 1
@@ -94,6 +94,7 @@ func TestCachedSettingsServerRestart(t *testing.T) {
 	closedts.TargetDuration.Override(ctx, &testServer.ClusterSettings().SV, 10*time.Millisecond*time.Duration(factor))
 	closedts.SideTransportCloseInterval.Override(ctx, &testServer.ClusterSettings().SV, 10*time.Millisecond*time.Duration(factor))
 
+	const expectedSettingsCount = 3
 	testutils.SucceedsSoon(t, func() error {
 		store, err := testServer.GetStores().(*kvserver.Stores).GetStore(1)
 		if err != nil {
@@ -103,10 +104,18 @@ func TestCachedSettingsServerRestart(t *testing.T) {
 		if err != nil {
 			return err
 		}
-		if len(settings) == 0 {
-			return errors.New("empty settings loaded from store")
+
+		// Previously, we checked if len(settings) > 0, which led to a race
+		// condition where, in rare cases (under --race), the settings watcher
+		// had not yet received some settings through rangefeed. If we exit this
+		// function and assign expectedSettingsCount with those incomplete
+		// settings, the settings watcher may receive the remaining settings
+		// before we stop the server. See issue #124419 for more details.
+		if len(settings) < expectedSettingsCount {
+			return errors.Newf("unexpected count of settings: expected %d, found %d",
+				expectedSettingsCount, len(settings))
 		}
-		settingsCache = settings
+		expectedSettingsCache = settings
 		return nil
 	})
 	testServer.Stopper().Stop(context.Background())
@@ -150,11 +159,11 @@ func TestCachedSettingsServerRestart(t *testing.T) {
 		if initialBoot {
 			return errors.New("server should not require initialization")
 		}
-		if !assert.ObjectsAreEqual(state.initialSettingsKVs, settingsCache) {
+		if !assert.ObjectsAreEqual(expectedSettingsCache, state.initialSettingsKVs) {
 			return errors.Newf(`initial state settings KVs does not match expected settings
 Expected: %+v
 Actual:   %+v
-`, settingsCache, state.initialSettingsKVs)
+`, expectedSettingsCache, state.initialSettingsKVs)
 		}
 		return nil
 	})

